include/votca/csg/beadmap.h

```diff
@@ -0,0 +1,235 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef VOTCA_CSG_BEADMAP_H
+#define VOTCA_CSG_BEADMAP_H
+
+#include "cgbeadstencil.h"
+#include "csgtopology.h"
+#include "molecule.h"
+#include <string>
+#include <vector>
+#include <votca/tools/property.h>
+#include <votca/tools/vec.h>
+
+namespace votca {
+namespace csg {
+
+class BoundaryCondition;
+
+/**
+ * @brief BeadMap for coarse graining an atomistic representation to a coarser
+ * representation
+ *
+ * This abstract interface class that sets the requirements for creating other
+ * classes that define how a set of atomistic beads could be used to update a
+ * coarse grained representation. For instance if one was to use H2O as an
+ * example we could create a single bead to determine how a water molecule
+ * behaves as opposed to using 3 separate atomic particles. Though this class
+ * does not create the coarse grained representation of water it does store the
+ * relationship between a coarse grained and atomistic representation.
+ *
+ * The relationship defined by the map allows one to update the positions,
+ * velocities, forces, masses and orientations of a coarse grained bead with
+ * its atomistic pieces.
+ *
+ * Atomistic Representation          Coarse Grained Representation
+ *
+ *  H - O                                   H2O
+ *      |
+ *      H
+ *
+ * Note that the bead map does not have ownership over either coarse grained
+ * representation or atomisitic representation it can only manipulate them
+ * when they are passed in, it does not and should not have responsibility for
+ * their memory managment.
+ */
+class BeadMap {
+ public:
+  virtual ~BeadMap(){};
+  virtual void Apply(const BoundaryCondition *boundaries,
+                     std::map<std::string, const Bead *> atomistic_beads,
+                     Bead *cg_bead) const = 0;
+
+  virtual void Initialize(const std::vector<std::string> subbeads,
+                          std::vector<double> weights,
+                          std::vector<double> ds) = 0;
+
+  void Initialize(std::vector<std::string> subbeads,
+                  std::vector<double> weights) {
+    std::vector<double> empty;
+    Initialize(subbeads, weights, empty);
+  }
+
+  std::vector<std::string> getAtomicBeadNames() const;
+
+  virtual std::unique_ptr<BeadMap> Clone() const = 0;
+
+ protected:
+  BeadMap(){};
+  struct element_t {
+    double weight_;
+    double force_weight_;
+  };
+  std::unordered_map<std::string, element_t> matrix_;
+  friend class AtomToCGMoleculeMapper;
+};
+
+/*******************************************************
+    Linear map for spherical beads
+*******************************************************/
+
+class Map_Sphere : public BeadMap {
+ public:
+  void Apply(const BoundaryCondition *boundaries,
```

Comment:
Can you use a more descriptive name than Apply. Why not have a return value, i.e the cg bead?